src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Instrument.common.cs

```diff
@@ -0,0 +1,50 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// Instrument{T} is the base class from which all non-observable instruments will inherit from.
+    /// </summary>
+    /// <remarks>
+    /// This class supports only the following generic parameter types: <see cref="byte" />, <see cref="short" />, <see cref="int" />, <see cref="long" />, <see cref="float" />, <see cref="double" />, and <see cref="decimal" />
+    /// </remarks>
+    public abstract partial class Instrument<T> : Instrument where T : struct
+    {
+        /// <summary>
+        /// Create the metrics instrument using the properties meter, name, description, and unit.
+        /// All classes extending Instrument{T} need to call this constructor when constructing object of the extended class.
+        /// </summary>
+        /// <param name="meter">The meter that created the instrument.</param>
+        /// <param name="name">The instrument name. cannot be null.</param>
+        /// <param name="unit">Optional instrument unit of measurements.</param>
+        /// <param name="description">Optional instrument description.</param>
+        protected Instrument(Meter meter, string name, string? unit, string? description) : base(meter, name, unit, description)
+        {
+            ValidateTypeParameter<T>();
+        }
+
+        /// <summary>
+        /// Record the measurement by notifying all <see cref="MeterListener" /> objects which listeneing to this instrument.
+        /// </summary>
+        /// <param name="measurement">The measurement value.</param>
+        protected void RecordMeasurement(T measurement) => RecordMeasurement(measurement, Instrument.EmptyTags.AsSpan());
+
+        /// <summary>
+        /// Record the measurement by notifying all <see cref="MeterListener" /> objects which listeneing to this instrument.
+        /// </summary>
+        /// <param name="measurement">The measurement value.</param>
+        /// <param name="tags">A span of key-value pair tags associated with the measurement.</param>
+        protected void RecordMeasurement(T measurement, ReadOnlySpan<KeyValuePair<string, object?>> tags)
+        {
+            LinkedListNode<ListenerSubscription>? current = _subscriptions.First;
```

Comment:
I didn't measure it and compare it with the prototype. I am not expecting the extra memory here is issue at all because it is proportional to the number of listeners which is small. Also, the prototype was reallocating a new array with every operation on the subscription which would cause more temporary allocation than the current implementation. Perf wise I expect the current implementation would be faster than using the arrays. The reason is arrays need to check the bounds every time you access any item while current implementation doesn't. I'll try to do some measurements after merging these changes and we can enhance more if needed or found any issue.

Reply:
I agree with @noahfalk. Arrays tend to have much better memory locality. The well-predicted bounds check is many times cheaper than a CPU cache miss.

> the reason is arrays need to check the bounds every time accessing any item while current implementation doesn't

Was the original prototype code accessing the array in a loop where the bounds-check would be optimized out?

Reply:
That is right. the original prototype was using the array in the look https://github.com/noahfalk/opentelemetry-dotnet/blob/80e6ff71dc44cea192c4f0ceb14e63a024c7d155/example3/Microsoft.Diagnostics.Metric/MeterInstrument.cs#L231 so it could be optimized out. As I mentioned in didn't compare both yet but this is something I'll do later outside this PR.

Reply:
> I agree with @noahfalk Noah Falk FTE. Arrays tend to have much better memory locality. The well-predicted bounds check is many times cheaper than a CPU cache miss.

Note that the list would be small any way as I am expecting in more regular cases, we'll have one or two items at most in the list.